### ISSUE TITLE
config.guess for newer architectures

### ISSUE THREAD

**Issue by lee218llnl** [LLNL/spack#2063]
@tgamblin @becker33 @mplegendre: as discussed, several packages contain an old config.guess which cannot detect some more modern architectures. It would be nice if Spack could replace the config.guess in those packages with a newer one.

**Comment by tgamblin** [comment]
`config.guess` is strictly an autotools thing, correct?  After #1186 is merged, we'll have a special package for autotools where we can consolidate this.  I think adding an option at the package base class level, e.g.:

```python
class AutotoolsPackage(Package):
    patch_config_guess = True

    def patch(self):
        # do default patching for config_guess if needed for self.spec.architecture
        # *maybe* query the platform for an up to date config.guess
        # Not sure if the new config.guess belongs with AutotoolsPackage or with the platform.
        # leaning towards AutotoolsPackage

class Foo(AutotoolsPackage):
    # can optionally set the patch option to False or override `patch()` entirely
```

Does that seem reasonable?  The other option would be to have autotools packages depend on autotools and require a new enough version for the host architecture.  Then we would just run `autoreconf`.  Patching `config.guess` seems simpler to me.

Thoughts?  @alalazo @mplegendre @lee218llnl

**Comment by lee218llnl** [comment]
@tgamblin While it'll be a bit of a pain to iterate through all the packages that need this, I think the AutoTools package would work for me. Looking at my bash history, here's what was needed for my STAT build on rzmanta:

  233  cp ~/Downloads/downloads/config.guess var/spack/stage/autoconf-2.69-uaaxtq2v2pnht4ofk7ldueqz3qghphvj/autoconf-2.69/build-aux/
  241  cp ~/Downloads/downloads/config.guess var/spack/stage/libelf-0.8.13-olmdmazsjo2frthnosgr635uj4se2sz4/libelf-0.8.13/
  303  cp ~/Downloads/downloads/config.guess var/spack/stage/launchmon-1.0.2-igfteri76v2ucf4ki6x6ggqvhwlnu7np/launchmon-v1.0.2/config/
  305  cp ~/Downloads/downloads/config.guess var/spack/stage/lzma-4.32.7-bdsgfyy4kkwrfz7yizpuibugqvywyn7u/lzma-4.32.7/
  315  cp ~/Downloads/downloads/config.guess var/spack/stage/launchmon-1.0.2-igfteri76v2ucf4ki6x6ggqvhwlnu7np/launchmon-v1.0.2/config
  318  cp ~/Downloads/downloads/config.guess ./
  319  cp config.guess var/spack/stage/libtiff-4.0.6-u6hkencvr5ykexilsxxhuxywo35vow4r/tiff-4.0.6/config/
  321  cp config.guess var/spack/stage/graphviz-2.38.0-dskek5hkkd57wbx65kkosmj5r6gecd5v/graphviz-2.38.0/config/
  331  cp config.guess var/spack/stage/graphviz-2.38.0-dskek5hkkd57wbx65kkosmj5r6gecd5v/graphviz-2.38.0/libltdl/config/
  337  cp config.guess var/spack/stage/fontconfig-2.11.1-emdqsy473iz23jyzh3pfknwi47gqn3nh/fontconfig-2.11.1/
  339  cp config.guess var/spack/stage/cairo-1.14.0-oetgsghf32iulpczyenxaexklnv5exzm/cairo-1.14.0/build/
  346   cp config.guess var/spack/stage/py-pygobject-2.28.3-cfxoyh26dbalihvmyex2ckr4aput5qc7/pygobject-2.28.3/
  350  cp config.guess var/spack/stage/py-pygtk-2.24.0-scira5jkebvrb366pbr7x5vpdzsdcnt7/pygtk-2.24.0/
  352  cp config.guess var/spack/stage/stat-3.0.0-kqld2njwx4w66kboa6xsuesgfkjuf7fd/STAT-3.0.0/config/


Note that because the config.guess file may not reside in the top level directory of a package, that we will have to search for the package's original config.guess file and that in some cases (i.e., graphviz), there may be a subpackage that requires the patch too.

As to where the updated config.guess should come from, perhaps we should start by checking for it in /usr/share/automake-X.Y? We can also see if the spack instance has an automake installed and use its config.guess.

**Comment by lee218llnl** [comment]
Ah, that's much easier. We still have the subpackage problem, though, but this isn't too tricky to resolve.

I'm OK with installing automake and using its config.guess for all autotools packages.

**Comment by lee218llnl** [comment]
FYI, I am working on this and think I have something that will work, but am waiting for our testbed to come back up so I can test it out. Stay tuned for a PR.

**Comment by alalazo** [comment]
@lee218llnl I'll close this as I think it was fixed in #2221
